core/src/main/java/software/amazon/awssdk/async/AsyncResponseHandler.java

```diff
@@ -46,6 +49,8 @@
      * will be treated as a failure of the response and the {@link #exceptionOccurred(Throwable)} callback will be invoked.
      * </p>
      *
+     * <p>This callback may never be called if the response has no content or if an error occurs.</p>
```

Comment:
So for non-streaming operations, AsyncResponseHandler#onStream will not be called since there won't be a customer provided response handler. SdkHttpResponseHandler#onStream will be called however as shown in the sync adapter.